front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/utils.ts

```diff
@@ -184,6 +184,33 @@ export const getSavedMacroNodes = async (
   return result;
 };
 
+/**
+ * Check if every new nodes (from import) are already in the DB (by path_item_key).
+ * If not we persist them.
+ */
+export const persistNodesFromFileImport = async (
+  state: MacroEditorState,
+  dispatch: AppDispatch
+) => {
+  const saved = await getSavedMacroNodes(state, dispatch);
+  const savedKeys = new Set(saved.map((n) => n.path_item_key));
+
+  const nodesFromFileImport = state.nodes.filter(
+    (n): n is NodeIndexed => n != null && !n.dbId && !savedKeys.has(n.path_item_key)
```

Comment:
I'd prefer to avoid `n is NodeIndexed` if we can, because it's a kind of cast. TypeScript should be smart enough to figure out that `state.nodes.filter((n) => n !== null)` returns an array of `NodeIndexed` without a type predicate. Maybe TypeScript gets confused by the more complicated condition, in which case I think it's fine to perform two `filter()`s.